File: sdks/node/src/session/adapters/anthropic.ts
/**
 * Adapters from Anthropic's wire format to the Weave Session SDK types.
 *
 * Use these when manually instrumenting calls to `client.messages.create`
 * (autopatched Anthropic integrations handle conversion automatically).
 *
 * Port of `weave/session/adapters/anthropic.py`.
 */

import {Usage} from '../types';

/**
 * Extract usage from an Anthropic Messages API `Message`.
 *
 * Anthropic types the cache fields as nullable; `null`/`undefined` is
 * treated as zero.
 */
export function usageFromAnthropic(message: Record<string, unknown>): Usage {
  const usage = (message.usage as Record<string, unknown> | undefined) ?? {};
  return new Usage({
    inputTokens: (usage.input_tokens as number | undefined) ?? 0,
    outputTokens: (usage.output_tokens as number | undefined) ?? 0,
    cacheCreationInputTokens:
      (usage.cache_creation_input_tokens as number | undefined) ?? 0,
    cacheReadInputTokens:
      (usage.cache_read_input_tokens as number | undefined) ?? 0,
  });
}

File: weave/conversation/adapters/anthropic.py
"""Adapters from Anthropic's wire format to the Weave Conversation SDK types.

Use these when manually instrumenting calls to ``client.messages.create``
(the autopatched Anthropic integration handles conversion automatically).
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from weave.conversation.types import Usage

if TYPE_CHECKING:
    from anthropic.types import Message as AnthropicMessage

__all__ = ["usage_from_anthropic"]


def usage_from_anthropic(message: AnthropicMessage) -> Usage:
    """Extract usage from an Anthropic Messages API ``Message``.

    Anthropic types the cache fields as ``Optional[int]``; ``None`` is
    equivalent to zero for our purposes.
    """
    usage = message.usage
    return Usage(
        input_tokens=usage.input_tokens,
        output_tokens=usage.output_tokens,
        cache_creation_input_tokens=usage.cache_creation_input_tokens or 0,
        cache_read_input_tokens=usage.cache_read_input_tokens or 0,
    )
